### PR TITLE
Adding support for view_frame

### PR DESCRIPTION
**Pull request by ahcorde**
Moving t2f_tools to ROS 2

 - Adding support for ROS 2.0

### CODE REVIEW

File: tf2_ros/src/tf2_ros/buffer.py [hunk 1]
```diff
@@ -67,7 +67,10 @@ def __init__(self, cache_time = None):
         self._callbacks_to_remove = []
         self._callbacks_lock = threading.RLock()
 
-    def __get_frames(self, req):
+        if debug:
+            self.srv = node.create_service(FrameGraph, 'tf2_frames', self.__get_frames)
+
+    def __get_frames(self, req, res):
```
Comment (dirk-thomas) [inline tf2_ros/src/tf2_ros/buffer.py:74]: This looks odd. The function gets a `res` argument but returns the response?

File: tf2_tools/CHANGELOG.rst [hunk 2]
```diff
@@ -2,6 +2,10 @@
 Changelog for package tf2_tools
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
+2.0.0 (2019-11-04)
+-------------------
+ahcorde: Adding support for ROS 2.0
+
```
Comment (dirk-thomas) [inline tf2_tools/CHANGELOG.rst]: Since this package hasn't been released yet it shouldn't have an entry in the changelog file yet (or only a section named `Forthcoming`).

File: tf2_tools/package.xml [hunk 3]
```diff
@@ -1,24 +1,30 @@
-<package>
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format2.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="2">
   <name>tf2_tools</name>
-  <version>0.9.1</version>
+  <version>2.0.0</version>
```
Comment (dirk-thomas) [inline tf2_tools/package.xml]: The version number must be the same as the other packages in this repo.

File: tf2_tools/scripts/view_frames.py [hunk 4]
```diff
@@ -28,54 +28,77 @@
...
+    while not cli.wait_for_service(timeout_sec=1.0):
+        node.get_logger().info('service not available, waiting again...')
+
+    future = cli.call_async(req)
+    rclpy.spin_until_future_complete(node, future)
+
+    try:
+        result = future.result()
+    except Exception as e:
+        node.get_logger().info('Service call failed %r' % (e,))
```
Comment (dirk-thomas) [inline tf2_tools/scripts/view_frames.py]: This should probably be an `error` instead of `info`. Also returning a non-zero return code would be good in this case.

File: tf2_tools/CMakeLists.txt [hunk 5]
```diff
@@ -1,16 +1,12 @@
...
 
-install(PROGRAMS scripts/view_frames.py
-        DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION}
+install(
+  PROGRAMS
+  scripts/view_frames.py
+  DESTINATION lib/${PROJECT_NAME}
 )
+
+ament_package()
```
Comment (dirk-thomas) [inline tf2_tools/CMakeLists.txt:12]: Maybe enable the common linters for this package?

File: tf2_ros/src/tf2_ros/buffer.py (hunk 1 above)
Comment (ahcorde) [reply tf2_ros/src/tf2_ros/buffer.py:74]: I'm using the same idea as in the ROS 2.0 [demos](https://github.com/ros2/demos/blob/master/demo_nodes_py/demo_nodes_py/services/add_two_ints_server.py#L27) or [examples](https://github.com/ros2/examples/blob/master/rclpy/services/minimal_service/examples_rclpy_minimal_service/service_member_function.py#L27). is it not need it to return the response?

File: tf2_tools/CMakeLists.txt (hunk 5 above)
Comment (dirk-thomas) [reply tf2_tools/CMakeLists.txt:12]: This was actually a bad advice. This will lead the more differences between the upstream repo and this fork which makes porting changes between them more effort. Please revert the added checks and the style updates - sorry for the hassle.

File: tf2_tools/CHANGELOG.rst (hunk 2 above)
Comment (dirk-thomas) [reply tf2_tools/CHANGELOG.rst]: Its probably best to remove the new changelog section. It will be auto-generated by `catkin_generate_changelog` before the next release based on commit messages.

File: tf2_tools/package.xml [hunk 6]
```diff
@@ -1,24 +1,30 @@
-<package>
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format2.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="2">
   <name>tf2_tools</name>
-  <version>0.9.1</version>
+  <version>0.12.1</version>
```
Comment (dirk-thomas) [inline tf2_tools/package.xml]: The versions in the repo have already been updated again to 0.12.4.

File: tf2_ros/src/tf2_ros/buffer.py (hunk 1 above)
Comment (dirk-thomas) [reply tf2_ros/src/tf2_ros/buffer.py:74]: I created a separate ticket for this (ros2/rclpy#464) so the patch can stay as is.

File: tf2_ros/src/tf2_ros/buffer.py [hunk 7]
```diff
@@ -1,9 +1,9 @@
 # Copyright (c) 2008, Willow Garage, Inc.
 # All rights reserved.
-# 
+#
```
Comment (dirk-thomas) [inline tf2_ros/src/tf2_ros/buffer.py]: Any unrelated whitespace changes should probably also be revert to reduce the difference to upstream as much as possible.

Comment (ahcorde) [reply tf2_ros/src/tf2_ros/buffer.py]: Atom is removing the whitespaces, solved

File: tf2_tools/scripts/view_frames.py [hunk 8]
```diff
@@ -28,32 +28,60 @@
...
-    with open('frames.gv', 'w') as f:
-        f.write(generate_dot(data))
-    subprocess.Popen('dot -Tpdf frames.gv -o frames.pdf'.split(' ')).communicate()
+    try:
+        result = future.result()
+    except Exception as e:
+        node.get_logger().error('Service call failed %r' % (e,))
+        cli.destroy()
+        node.destroy_node()
+        rclpy.shutdown()
```
Comment (dirk-thomas) [inline tf2_tools/scripts/view_frames.py]: How about moving the `destroy` / `shutdown` calls into a `finally:` block to be triggered in this exception case as well as on normal return to avoid duplication?

File: tf2_ros/src/tf2_ros/buffer.py [hunk 9]
```diff
@@ -51,7 +51,7 @@ class Buffer(tf2.BufferCore, tf2_ros.BufferInterface):
     known frames. 
     """
 
-    def __init__(self, cache_time = None):
+    def __init__(self, cache_time = None, debug = False, node = None):
```
Comment (dirk-thomas) [inline tf2_ros/src/tf2_ros/buffer.py]: Since both new arguments can only be used together how about just passing `node` conditionally when debugging is desired?

Please update the docblock of the constructor accordingly.

File: tf2_tools/scripts/view_frames.py [hunk 10]
```diff
@@ -28,32 +28,60 @@
...
+            'Result:'+ str(result) )
+        data = yaml.load(result.frame_yaml)
+        with open('frames.gv', 'w') as f:
+           f.write(generate_dot(data, node))
+        subprocess.Popen('dot -Tpdf frames.gv -o frames.pdf'.split(' ')).communicate()
+    node.destroy_node()
+    rclpy.shutdown()
 
-def generate_dot(data):
+def generate_dot(data, node):
```
Comment (dirk-thomas) [inline tf2_tools/scripts/view_frames.py]: How about passing a clock into this function instead - similar to #198?

File: tf2_tools/scripts/view_frames.py [hunk 11]
```diff
@@ -72,7 +100,7 @@ def generate_dot(data):
             root = map['parent']
     dot += 'edge [style=invis];\n'
     dot += ' subgraph cluster_legend { style=bold; color=black; label ="view_frames Result";\n'
-    dot += '"Recorded at time: '+str(rospy.Time.now().to_sec())+'"[ shape=plaintext ] ;\n'
+    dot += '"Recorded at time: '+str(node.get_clock().now())+'"[ shape=plaintext ] ;\n'
```
Comment (dirk-thomas) [inline tf2_tools/scripts/view_frames.py]: Maybe format this as a float number in seconds instead? Currently it seems to be a tuple?

File: tf2_ros/src/tf2_ros/buffer.py [hunk 12]
```diff
@@ -67,7 +68,10 @@ def __init__(self, cache_time = None):
         self._callbacks_to_remove = []
         self._callbacks_lock = threading.RLock()
 
-    def __get_frames(self, req):
+        if node != None:
```
Comment (dirk-thomas) [inline tf2_ros/src/tf2_ros/buffer.py]: Should be `node is not None`.

**Review by dirk-thomas — APPROVED** [review]
Just one minor nitpick.

File: tf2_ros/src/tf2_ros/buffer.py (hunk 12 above)
Comment (ahcorde) [reply tf2_ros/src/tf2_ros/buffer.py]: Sure ;)

**Comment by dirk-thomas** [comment]
@ahcorde Ready to be merged? Should this be backported to distros before Foxy?

**Comment by dirk-thomas** [comment]
The shebang line of the `view_frames.py` script (https://github.com/ros2/geometry2/pull/192/files#diff-caff1681fadbfe3ebe6e97b4113d7916R1) wasn't updated. Therefore the script is being run with Python 2 and fails with a `SyntaxError`.

@ahcorde Please update the shebang line to `#!/usr/bin/env python3` and also backport that fix.

**Comment by ahcorde** [comment]
Fix: https://github.com/ros2/geometry2/pull/226 
Eloquent backport: https://github.com/ros2/geometry2/pull/227
